exp/runtime/internal/controllers/extension_controller.go

```diff
@@ -21,74 +21,169 @@ import (
-	r.Registry.RemoveRuntimeExtension(ext)
+func patchExtension(ctx context.Context, patchHelper *patch.Helper, ext *runtimev1.Extension, options ...patch.Option) error {
+	// Can add errors and waits here to the Extension object.
+	return patchHelper.Patch(ctx, ext, options...)
+}
 
+func (r *Reconciler) reconcileDelete(_ context.Context, extension *runtimev1.Extension) (ctrl.Result, error) {
+	// FIXME: This isn't implemented on the client/registry.
+	if err := r.RuntimeClient.Extension(extension).Unregister(); err != nil {
+		return ctrl.Result{}, err
+	}
 	return ctrl.Result{}, nil
 }
 
-func (r *ExtensionReconciler) reconcile(ctx context.Context, ext *expv1.Extension) (ctrl.Result, error) {
-	// FIXME:
-	// * should call Discover on RuntimeExtension and then update Status
-	// Note: Has to work with ext.Spec.ClientConfig without underlying registry
-	// Q: Is it enough to do the Discovery only once initially or on each reconcile
+func (r *Reconciler) reconcile(ctx context.Context, extension *runtimev1.Extension) (ctrl.Result, error) {
+	if err := r.discoverRuntimeExtensions(ctx, extension); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
 
-	r.Registry.RegisterRuntimeExtension(ext)
+// FIXME: should this method use an api reader to ensure we don't hit the cache?
+func (r *Reconciler) syncRegistry(ctx context.Context) (reterr error) {
+	extensionList := runtimev1.ExtensionList{}
+	if err := r.Client.List(ctx, &extensionList); err != nil {
+		return err
+	}
+	var errs []error
+	for _, ext := range extensionList.Items {
+		extension := ext.DeepCopy()
+		// Initialize the patch helper for this extension.
+		patchHelper, err := patch.NewHelper(extension, r.Client)
+		if err != nil {
+			return err
+		}
+		if err := r.discoverRuntimeExtensions(ctx, extension); err != nil {
+			errs = append(errs, err)
+		} else {
+			conditions.MarkTrue(extension, runtimev1.RuntimeExtensionDiscovered)
+		}
 
-	return ctrl.Result{}, nil
+		defer func(extension runtimev1.Extension) { //nolint:gocritic
+			// Always attempt to patch the object and status after each reconciliation.
+			// Patch ObservedGeneration only if the reconciliation completed successfully
+			patchOpts := []patch.Option{}
+			if reterr == nil {
+				patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
+			}
+			if err = patchExtension(ctx, patchHelper, &extension, patchOpts...); err != nil {
+				reterr = kerrors.NewAggregate([]error{reterr, err})
+			}
+		}(*extension)
+	}
+
+	if len(errs) != 0 {
+		return kerrors.NewAggregate(errs)
+	}
+	// If all extensions are listed and discovered successfully the reconciler is now ready.
+	r.RuntimeClient.SetRegistryReady(true)
+	return nil
+}
+
+func (r Reconciler) discoverRuntimeExtensions(_ context.Context, extension *runtimev1.Extension) error {
+	extensions, err := r.RuntimeClient.Extension(extension).Discover()
```

Comment:
We will have to call this differently based on current registry design. Let's do a Zoom call and play around with it together a bit.

Reply:
Will come back to this when we have a client API for this call.